scripts/dts/python-devicetree/src/devicetree/edtlib.py

```diff
@@ -336,6 +339,64 @@ def _merge_includes(self, raw: dict, binding_path: Optional[str]) -> dict:
 
         return raw
 
+    def _merge_matches(self) -> None:
+        # Helper to merge properties matching another pattern property. For
+        # example, an included file may contain the pattern property
+        # '.*-supply$', and the binding may contain a matching regular property
+        # 'vin-supply' to constrain a specific matching property.
+        #
+        # The following rules apply:
+        #   - The type is always taken from the pattern property.
+        #   - If pattern property is deprecated, any matching property is
+        #     deprecated as well.
+        #   - Any other attributes are taken from the pattern property if not
+        #     specified in the regular property.
+
+        all_props = self.raw.get("properties")
+        if not all_props:
+            return
+
+        pprops = {}
+        rprops = {}
+        for name, prop in all_props.items():
+            if prop.get("pattern"):
+                pprops[name] = prop
+            else:
+                rprops[name] = prop
+
+        for rname, rprop in rprops.items():
+            for pname, pprop in pprops.items():
+                if re.match(pname, rname):
+                    if (rprop.get("type") and
+                        rprop.get("type") != pprop.get("type")):
+                        _err(f"Property {rname} in {self.path} matches {pname} "
+                             f"pattern, but types do not match")
+
+                    if (rprop.get("deprecated") is not None and
+                        pprop.get("deprecated") and not rprop.get("deprecated")):
+                        _err(f"Property {rname} in {self.path} matches {pname} "
+                             f"pattern, deprecated can not be undone")
```

Comment:
This can be followed up in another PR, but should this be an error in general, beyond pattern properties? Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/1442a6cf5bfe0fa3d486587c28af97f172649a6c/doc/build/dts/bindings-syntax.rst?plain=1#L327-L328

Same with `required: false`:
https://github.com/zephyrproject-rtos/zephyr/blob/1442a6cf5bfe0fa3d486587c28af97f172649a6c/doc/build/dts/bindings-syntax.rst?plain=1#L252-L254
https://github.com/zephyrproject-rtos/zephyr/blob/1442a6cf5bfe0fa3d486587c28af97f172649a6c/doc/build/dts/bindings-syntax.rst?plain=1#L711-L713

cc @mbolivar-ampere